### PR TITLE
[Sol - Yul] Add support for ABIDecode.

### PR DESCRIPTION
**Pull request by aarlt**
Part of #8343.

- [X] basic working version, decoding of data from memory & call-data
- [X] `test/libsolidity/semanticTests/abiencodedecode/abi_decode_simple_storage.sol`
    - `solUnimplementedAssert` in `libsolidity/codegen/YulUtilFunctions.cpp(1363)`
- [X] `test/libsolidity/semanticTests/abiEncoderV1/decode_slice.sol`
    - `invalid IR` generation of multiple definitions of same variables fixed
- [X] `test/libsolidity/semanticTests/abiEncoderV1/abi_decode_v2.sol`
    - `UnimplementedFeatureError` in `git/solidity/libsolidity/codegen/ir/IRGeneratorForStatements.cpp(1473)`

### CODE REVIEW

File: libsolidity/codegen/ir/IRGeneratorForStatements.cpp [hunk 1]
```diff
@@ -875,6 +875,61 @@ void IRGeneratorForStatements::endVisit(FunctionCall const& _functionCall)
 		m_code << templ.render();
 		break;
 	}
+	case FunctionType::Kind::ABIDecode:
+	{
+		Whiskers templ(R"(
+			let <retVars> := <abiDecode>(<offset>, add(<offset>, add(<offset>, <length>)))
```
Comment (chriseth) [inline libsolidity/codegen/ir/IRGeneratorForStatements.cpp]:
```suggestion
			let <retVars> := <abiDecode>(<offset>, add(<offset>, <length>))
```

File: libsolidity/codegen/ir/IRGeneratorForStatements.cpp [hunk 2]
```diff
@@ -875,6 +875,61 @@ void IRGeneratorForStatements::endVisit(FunctionCall const& _functionCall)
...
+		)");
+
+		TypePointer firstArgType = arguments.front()->annotation().type;
```
Comment (chriseth) [inline libsolidity/codegen/ir/IRGeneratorForStatements.cpp:893]:
```suggestion
		Type const* arrayType = arguments.front()->annotation().type;
```
Or `dataType`? or `encodedType`?

File: libsolidity/codegen/ir/IRGeneratorForStatements.cpp [hunk 3]
```diff
@@ -875,6 +875,61 @@ void IRGeneratorForStatements::endVisit(FunctionCall const& _functionCall)
...
+		else
+		{
+//			solUnimplementedAssert(false, "");
+			// memory
+			IRVariable var = convert(*arguments[0], *TypeProvider::bytesMemory());
+//			std::cout << "- from memory: " << var.type().toString() << " @ " << var.type().memoryHeadSize() << std::endl;
+//			std::cout << "- from memory: " << firstArgType->toString() << " @ " << firstArgType->memoryHeadSize() << std::endl;
+//			std::cout << "- from memory: " << referenceType->toString() << " @ " << referenceType->memoryHeadSize() << std::endl;
+			for (auto& slot : var.stackSlots())
+				std::cout << " -> " << slot << std::endl;
```
Comment (chriseth) [inline libsolidity/codegen/ir/IRGeneratorForStatements.cpp]: Look at `ArrayType::makeStackItems` to check the stack slots

**Comment by aarlt** [comment]
See current output here: https://gist.github.com/aarlt/a25642308fdda03ce28d54a0739a50f1

File: libsolidity/codegen/ir/IRGeneratorForStatements.cpp (hunk 3 above)
Comment (aarlt) [reply libsolidity/codegen/ir/IRGeneratorForStatements.cpp]: nice 👍

**Comment by aarlt** [comment]
There are still some problems with some `abi.decode` related tests, e.g.:

- `test/libsolidity/semanticTests/abiencodedecode/abi_decode_simple_storage.sol` fails in `libsolidity/codegen/YulUtilFunctions.cpp:1363` because of `solUnimplementedAssert`:
```c++
		else
		{
			if (_type.category() == Type::Category::Array)
				solUnimplementedAssert(false, "");  // <<<<<<<<<<<
			else if (_type.category() == Type::Category::Struct)
				solUnimplementedAssert(false, "");
			else
				solAssert(false, "Invalid non-value type for assignment.");
		}
```

- `test/libsolidity/semanticTests/abiEncoderV1/decode_slice.sol` fails because of invalid IR.
```
Invalid IR generated:
:137:17: Error: Variable name expr_22_offset already taken in this scope.
                let expr_22_offset := 0
                ^---------------------^
:138:17: Error: Variable name expr_22_length already taken in this scope.
                let expr_22_length := calldatasize()
                ^----------------------------------^
:149:17: Error: Variable name expr_37_offset already taken in this scope.
                let expr_37_offset := 0
                ^---------------------^
:150:17: Error: Variable name expr_37_length already taken in this scope.
                let expr_37_length := calldatasize()
                ^----------------------------------^
:162:17: Error: Variable name expr_53_offset already taken in this scope.
                let expr_53_offset := 0
                ^---------------------^
:163:17: Error: Variable name expr_53_length already taken in this scope.
                let expr_53_length := calldatasize()
                ^----------------------------------^
```
- `test/libsolidity/semanticTests/abiEncoderV1/abi_decode_v2.sol` fails because of `solAssert(members[i]->dataStoredIn(DataLocation::Memory), "");` in `libsolidity/codegen/YulUtilFunctions.cpp:1670`
```c++
		vector<map<string, string>> memberParams(members.size());
		for (size_t i = 0; i < members.size(); ++i)
		{
			solAssert(members[i]->memoryHeadSize() == 32, "");
			solAssert(members[i]->dataStoredIn(DataLocation::Memory), "");
			memberParams[i]["zeroValue"] = zeroValueFunction(*members[i], false);
		}
```

File: libsolidity/codegen/ir/IRGeneratorForStatements.cpp [hunk 4]
```diff
@@ -875,6 +875,45 @@ void IRGeneratorForStatements::endVisit(FunctionCall const& _functionCall)
 		m_code << templ.render();
 		break;
 	}
+	case FunctionType::Kind::ABIDecode:
+	{
+		Whiskers templ(R"(
+			let <retVars> := <abiDecode>(<offset>, add(<offset>, <length>))
```
Comment (chriseth) [inline libsolidity/codegen/ir/IRGeneratorForStatements.cpp]:
```suggestion
			<?+retVars>let <retVars> := </+retVars><abiDecode>(<offset>, add(<offset>, <length>))
```

File: libsolidity/codegen/ir/IRGeneratorForStatements.cpp [hunk 5]
```diff
@@ -875,6 +875,45 @@ void IRGeneratorForStatements::endVisit(FunctionCall const& _functionCall)
...
+			templ("length", var.part("length").name());
+		}
+		else
+		{
+			IRVariable var = convert(*arguments[0], *TypeProvider::bytesMemory());
+			templ("abiDecode", m_context.abiFunctions().tupleDecoder(targetTypes, true));
+			templ("offset", "add(" + var.part("mpos").name() + ", 32)");
+			templ(
+				"length",
+				m_utils.arrayLengthFunction(*TypeProvider::bytesMemory()) + "(" + var.part("mpos").name() + ")");
```
Comment (chriseth) [inline libsolidity/codegen/ir/IRGeneratorForStatements.cpp]:
```suggestion
				m_utils.arrayLengthFunction(*TypeProvider::bytesMemory()) + "(" + var.part("mpos").name() + ")"
			);
```

**Comment by chriseth** [comment]
The first problem is fine, it is not yet implemented. The second should be investigated in the generated code and the third will be fixed by this one: https://github.com/ethereum/solidity/pull/9020

**Comment by aarlt** [comment]
> The first problem is fine, it is not yet implemented. The second should be investigated in the generated code and the third will be fixed by this one: #9020

I will look into the second problem. The third one seem to be "fixed", but hits another `UnimplementedFeatureError`.

**Review by chriseth — APPROVED** [review]

**Pull request merged**